compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULayoutAnalysisAndDistribution.cpp

```diff
@@ -0,0 +1,517 @@
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/NVGPU/IR/NVGPUDialect.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace mlir::iree_compiler {
+
+namespace {
+
+static constexpr int maxTensorDims = 2;
+namespace DimType {
+static constexpr int Batch0 = 0;  // Batch dimension for tensor dim 0
+static constexpr int Batch1 = 1;  // Batch dimension for tensor dim 1
+static constexpr int LaneIdZ = 2;
+static constexpr int LaneIdY = 3;
+static constexpr int LaneIdX = 4;
+static constexpr int VecIdZ = 5;
+static constexpr int VecIdY = 6;
+static constexpr int VecIdX = 7;
+static constexpr int NumDims = 8;
+}  // namespace DimType
+
+static std::string typeToString(int i) {
+  switch (i) {
+    case DimType::Batch0:
+      return "Batch0";
+    case DimType::Batch1:
+      return "Batch1";
+    case DimType::LaneIdZ:
+      return "LaneIdZ";
+    case DimType::LaneIdY:
+      return "LaneIdY";
+    case DimType::LaneIdX:
+      return "LaneIdX";
+    case DimType::VecIdZ:
+      return "VecIdZ";
+    case DimType::VecIdY:
+      return "VecIdY";
+    case DimType::VecIdX:
+      return "VecIdX";
+    default:
+      return "";
+  }
+}
+
+struct Dimension {
+  int type;
+  int value;
+};
+
+using DimOrderArray = std::array<std::array<Dimension, 3>, maxTensorDims>;
+using OrderArray = std::array<std::array<int, 4>, maxTensorDims>;
+using DimArray = std::array<int, DimType::NumDims>;
+
+struct Layout {
+  Layout(const DimOrderArray &orders,
+         const std::array<int, maxTensorDims> &canonicalShape);
+  // Reduces the layout along the tensor dim i
+  void reduceDim(int i);
+  // Updates the batch dims of the layout given the tensor dims
+  void updateBatchDims(int dim0, int dim1);
+  // Computes the ith dimension expression for a given state
+  AffineExpr computeDim(int i, const DimArray &state, OpBuilder &builder);
+  bool operator==(const Layout &layout) const { return shape == layout.shape; }
+  bool operator!=(const Layout &layout) const { return shape != layout.shape; }
+  std::string str() const;
+
+  // Contains the shape of the layout
+  DimArray shape;
+  // Contains the order of layout dims for each of the tensor dims
+  OrderArray order;
+  // Shape of the tensor when used in mma
+  std::array<int, maxTensorDims> canonicalShape;
```

Comment:
is it meant to be maxVectorDim? There should be no more tensor when doing this conversion?

Reply:
So for the mma.m16n8k16 case, the A-matrix has a canonical shape 16x16, B has 16x8 and C has 16x8. So that's what I was referring to here as maxTensorDims (=2). But maybe this is confusing, should I rename maxTensorDims to maxVectorDims?